rust/connlib/snownet/src/node/allocations.rs

```diff
@@ -0,0 +1,234 @@
+use std::{
+    collections::{BTreeMap, btree_map::Entry},
+    fmt,
+    net::{IpAddr, SocketAddr},
+    time::Instant,
+};
+
+use bufferpool::BufferPool;
+use itertools::Itertools as _;
+use rand::{Rng, seq::IteratorRandom as _};
+use ringbuffer::{AllocRingBuffer, RingBuffer};
+use str0m::Candidate;
+use stun_codec::rfc5389::attributes::{Realm, Username};
+
+use crate::{
+    RelaySocket, Transmit,
+    allocation::{self, Allocation},
+    node::SessionId,
+};
+
+pub(crate) struct Allocations<RId> {
+    inner: BTreeMap<RId, Allocation>,
+    previous_relays_by_ip: AllocRingBuffer<IpAddr>,
+
+    buffer_pool: BufferPool<Vec<u8>>,
+}
+
+impl<RId> Allocations<RId>
+where
+    RId: Ord + fmt::Display + Copy,
+{
+    pub(crate) fn clear(&mut self) {
+        self.inner.clear();
```

Comment:
I guess it depends on how/when you plan to use `clear()` but perhaps it would be worth to keep track of the cleared ones in `previous_relays_by_ip`? 

something like:

```suggestion
        for allocation in self.inner.values() {
            self.previous_relays_by_ip
                .extend(server_addresses(allocation));
        }
        self.inner.clear();
```

Reply:
Good catch, thank you! I should write tests for this :sweat_smile:

Reply:
Now unit tested!